### ISSUE TITLE
patch_all() can patch modules multiple times, causing RecursionErrors

### ISSUE THREAD

**Issue by nathanreynolds-kcom** [aws/aws-xray-sdk-python#99]
We hit an issue where patch_all() was used in a Lambda function - after the 1000th call we saw a RecursionError when calling `requests.get()`.

Example:

```python
from aws_xray_sdk.core import patch_all
import requests
import sys

def lambda_handler(event, _context):
    for i in range(sys.getrecursionlimit()):
        patch_all()
    requests.get('https://google.com') # Triggers RecursionError
```

We should have called `patch_all()` outside of the function, but this still feels like a gotcha.

`core.patcher._patch` has this code:

```python
def _patch(module_to_patch):
    # [...]
    if module_to_patch in _PATCHED_MODULES:
        log.debug('%s already patched', module_to_patch)
```

So it recognises the multiple patch, but repatches anyway.

Should `_patch` return early if it recognises the patch is already in place?

**Comment by haotianw465** [comment]
Thank you for reporting the issue. This is a bug and we will work on a fix soon. Just want to clarify, does calling `patch_all()` outside the handler workaround the problem for you? When you say `the 1000th call` does it mean you got this error in your 1000th lambda function execution?

**Comment by haotianw465** [comment]
The issue is fixed in 2.2.0 https://pypi.org/project/aws-xray-sdk/. Closing.
